test/lib/PortMap.h

```diff
@@ -0,0 +1,84 @@
+// Track TCP ports to use with the tester.
+class PortMap {
+  public:
+
+    static const int64_t START_PORT = 10000;
+    static const int64_t MAX_PORT = 20000;
+
+    // Get an unused port
+    uint16_t getPort()
+    {
+        lock_guard<mutex> lock(_m);
+        uint16_t port = 0;
+
+        // Reuse the first available freed port.
+        if (_returned.size()) {
+            port = *_returned.begin();
+            if (waitForPort(port) == 0) {
+                _returned.erase(_returned.begin());
+                return port;
+            }
+        }
+
+        // Otherwise, grab from the next port in our range.
+        do {
+            port = _from;
+            _from++;
+            if (waitForPort(port) == 0) {
+                return port;
+            }
+        } while (port<=MAX_PORT);
+
+        // Should never reach this.
+        cout << "Ran out of available ports after port " << port << endl;
+        SASSERT(port<MAX_PORT);
+        return 1;
+    }
+
+    // Free up previously used port to be used again
+    void returnPort(uint16_t port)
+    {
+        lock_guard<mutex> lock(_m);
+        _returned.insert(port);
+    }
+
+    // Waits for a particular port to be free to bind to. This is useful when we've killed a server, because sometimes
+    // it takes the OS a few seconds to make the port available again.
+    int waitForPort(uint16_t port) {
```

Comment:
why is this an  int instead of a bool? only ever returns 0 or 1

Reply:
I kept it from the previous implementation. I assume it's an int because the bind function returns 0 for success and 1 for failure, which is the opposite of the truth value of those. It could be done, just seemed non-trivial to replace the logic in BedrockTester that relies on this. I agree, it's confusing though.